### PR TITLE
Micromegas decoder update

### PR DESCRIPTION
**Pull request by hupereir**
[comment]: <> (Please tell us something about this pull request)

This is TPOT fix for overlapping time frames, similar to https://github.com/sPHENIX-Collaboration/coresoftware/pull/4292
TPOT decoder still uses taggers (GTM and enddat) to associate FEE data to 40bits gtm BCO, unlike the TPC, but this is really an orthogonal change. 

## Types of changes
[comment]: <> ( What types of changes does your code introduce? Put an `x` in all the boxes that apply: )
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work for users)
- [ ] Requiring change in macros repository (Please provide links to the macros pull request in the last section)
- [x] I am a member of [GitHub organization of sPHENIX Collaboration](https://github.com/orgs/sPHENIX-Collaboration/people), EIC, or ECCE (contact Chris Pinkenburg to join)

## What kind of change does this PR introduce? (Bug fix, feature, ...)

[comment]: <> ( What does this PR do? Linking to talk in software meeting encouraged )

## TODOs (if applicable)

[comment]: <> ( In case this is a draft PR, e.g. for running checks using Jenkins, please make the pull request as a draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/  )

### Motivation / context
TPOT Micromegas decoding could mis-associate FEE data to neighboring 40-bit GTM BCOs when time frames overlap. This PR improves the rollover-aware GTM↔FEE BCO matching and updates the streaming/pool filling workflow so evaluation and (optional) truncated/overlapping waveform recovery are driven by a target BCO.

### Key changes
- **Rollover-aware BCO matching utilities + tuning**
  - Added signed/unsigned clock-difference helpers and updated GTM↔FEE matching logic to use them consistently.
  - Updated matching parameters: **`m_max_gtm_bco_diff = 60`** and **`m_max_fee_sync_time = 1024 * 96`**.
- **Target-BCO driven streaming/pool filling**
  - `FillPool(...)` API changed to `FillPool(const uint64_t target_bco)` and is now filled in a loop controlled by `is_more_data_required(target_bco)`.
  - After pool filling, it calls `fill_evaluation_tree(target_bco)` and (if enabled) `recover_truncated_waveforms(target_bco)`.
  - In `FillMicromegasPool()` the streaming manager computes/passes a **`ref_bco_minus_range`** into micromegas `FillPool(...)`.
- **Waveform recovery for truncated/overlapping channels**
  - Implemented `recover_truncated_waveforms(target_bco)` to scan for an overlapping BCO, create shifted/truncated waveform copies for missing channels, and reinsert recovered hits for downstream use.
  - Added an enable/disable flag: `m_recover_truncated_waveforms` / `SetRecoverTruncatedWaveforms(...)`.
- **Micromegas raw-hit interface + minor fixes**
  - `MicromegasRawHitv3` now exposes `get_adc_waveforms()` for waveform access.
  - Minor messaging cleanup in `MicromegasRawHitv3.cc` (typo in warning text).

### Potential risk areas
- **Reconstruction behavior changes:** BCO association rules/tolerances and the reference computation path were modified, which can change which BCOs/waveforms are treated as matched and/or recovered.
- **Streaming/pool workflow changes:** `FillPool` signature and the “fill → evaluate → recover” sequencing are changed; verify downstream expectations for evaluation-tree fields and stored raw-hit bookkeeping.
- **Interface/data-structure impacts:** SingleMicromegas pool internals were refactored (including raw-hit storage patterns) and `MicromegasRawHitv3` added a public accessor—confirm consumer compatibility.
- **Performance:** waveform recovery adds additional scanning/copying work for overlapping/truncated scenarios; validate timing/CPU under high occupancy.

### Possible future improvements
- Add focused regression tests specifically covering **rollover boundaries** and **overlapping time-frame recovery** (including negative/offset adjustments between Tagger and GL1/target BCO path).
- Profile recovery scanning/allocation patterns and reduce copying where possible.
- Consolidate and document the signed/unsigned diff conventions to reduce future drift across decoders.
- AI can make mistakes—please validate against the matching/recovery code paths and relevant QA outputs before approving.

### CODE REVIEW

**Comment by hupereir** [comment]
Screenshot_20260611_152259

File: offline/framework/fun4allraw/SingleMicromegasPoolInput_v2.cc [hunk 1]
```diff
@@ -1022,13 +980,168 @@ void SingleMicromegasPoolInput_v2::process_fee_data(int packet_id, unsigned int
...
+            target = new MicromegasRawHit_impl;
+
+            // copy relevant members from source
+            target->set_bco(target_fee_bco);
+            target->set_gtm_bco(source->get_gtm_bco());
+            target->set_packetid(source->get_packetid());
+            target->set_fee(source->get_fee());
+            target->set_channel(source->get_channel());
+            target->set_sampaaddress(source->get_sampaaddress());
+            target->set_sampachannel(source->get_sampachannel());
```
[the remark this answers is not in the document]
Comment (hupereir) [reply offline/framework/fun4allraw/SingleMicromegasPoolInput_v2.cc]: This is intended. target_bco, coming from GL1, is actually expected to match the FEE BCO better than found_bco. Hence the choice of using the former to generate a new hit.

File: offline/framework/fun4allraw/MicromegasBcoMatchingInformation_v2.cc [hunk 2]
```diff
@@ -132,6 +137,51 @@ unsigned int MicromegasBcoMatchingInformation_v2::m_max_multiplier_adjustment_co
...
+  const auto diff = get_signed_gtm_bco_diff(first, second);
+  return uint64_t((diff<0) ? -diff:diff);
+}
+
+//___________________________________________________
+uint32_t MicromegasBcoMatchingInformation_v2::get_unsigned_fee_bco_diff(uint32_t first, uint32_t second)
+{
+  const auto diff = get_signed_fee_bco_diff(first, second);
+  return uint32_t((diff<0) ? -diff:diff);
+}
```
[the remark this answers is not in the document]
Comment (hupereir) [reply offline/framework/fun4allraw/MicromegasBcoMatchingInformation_v2.cc:181]: addressed with [820ad8a](https://github.com/sPHENIX-Collaboration/coresoftware/pull/4297/commits/820ad8a3b0a3d82fc1023177660ee072d82270bd)

**Comment by hupereir** [comment]
None of the Clang-tidy errors are related to this commit. 
The Varlgrind "unstable" is worrysome, but runs from a macro that do not execute this code (this is a simulation macro) The valgrind result on the Streaming reconstruction is green.
The last batch of commits is about cleanup and consolidation.
The code ran on 4 full runs without issue, namely 79527, 79564, 80279 and 80287

File: offline/framework/fun4allraw/SingleMicromegasPoolInput_v2.cc [hunk 3]
```diff
@@ -1022,13 +938,237 @@ void SingleMicromegasPoolInput_v2::process_fee_data(int packet_id, unsigned int
...
+    { continue; }
+
+    m_waveform.packet_id = m_fee_packet[fee];
+    m_waveform.fee_id = fee;
+    m_waveform.gtm_bco_first = bco_matching_information.get_bco_matching_reference().second;
+    m_waveform.fee_bco_first = bco_matching_information.get_bco_matching_reference().first;
+
+    // assign target bco and prediction
+    m_waveform.gtm_bco_gl1 = target_bco;
+    m_waveform.fee_bco_predicted_gl1 = bco_matching_information.get_predicted_fee_bco(target_bco).value();
```
[the remark this answers is not in the document]
Comment (hupereir) [reply offline/framework/fun4allraw/SingleMicromegasPoolInput_v2.cc]: this is irrelevant. Fill_evaluation_tree is only called once the BCOMatching objects are validated and thus the result of get_predicted_fee_bco is always valid.

**Comment by hupereir** [comment]
I'm done. This one is good to go.

**Comment by osbornjd** [comment]
Thanks for the confirmation - it looks like the memory leaks probably snuck in with the Acts update. I will investigate further

**Comment by osbornjd** [comment]
Git must be down as it is not letting me merge this due to "unable to read response from the server"...?

**Comment by pinkenburg** [comment]
The leak is from Acts - I tried to add a suppression for it but it somehow doesn't work. Need to look into this. It is static - so not critical to track down.

**Comment by hupereir** [comment]
I failed to notice this: 
image

I believe this is expected and a direct consequence of recovering the truncated waveforms. The reference histograms should be updated.
